provider/aws: resource ElastiCache

tmtk75 opened the pull request:
Hello everyone,

I implemented three AWS resources for ElastiCache.
* New resources for ElastiCache
  * `aws_ecache`
  * `aws_ecache_subnet_group`
  * `aws_ecache_security_group`
* Acceptance test summary
  * Launch clusters in both VPC & none-VPC
  * Ensure `memcached` and `redis` engine work
  * The acceptance test of aws_ecache takes about 12 minutes.
```
--- PASS: TestAccAWSElasticache (720.46s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    720.470s
```

It would be great for me if you reviewed, gave me feedbacks and then merged.

pmoust commented on builtin/providers/aws/resource_aws_ecache.go:
```diff
@@ -0,0 +1,256 @@
+	"time"
+
+	"github.com/hashicorp/aws-sdk-go/aws"
+	elasticcache "github.com/hashicorp/aws-sdk-go/gen/elasticache"
```
seems there is a naming inconsistency across this PR. `ecache`, `elasticcache`, `elasticache`.
Let's pick the latter.

tmtk75 replied:
Thanks for checking. I'm going to fix this inconsistency if we concluded discussion for naming.

This `elasticcache` alias in import is not needed to compile. This actually is due to my personal reason. My editor doesn't work well for its jump feature when jumping to `aws-sdk-go/aws`. It was hard to find interface of API during implementation if there was no this prefix, so I put it.

But now I can eliminate such prefixes and I'll do after the review.

radeksimko commented:
Out of curiosity, why `aws_ecache` and not `aws_elastic_cache`? I'd personally like to see the longer version as that's more descriptive and the DSL should be descriptive in the first place, that's IMO one of the reasons why people use it instead of JSON (CloudFormation) or Python (Trophosphere).

mitchellh commented:
I agree with @radeksimko on the name, we'll do a more full review soon!

tmtk75 commented:
@radeksimko @mitchellh Thanks for your feedback.
Actually I was concerned about which name, short or full, I should use. Then I just followed other resources which have `elastic` prefix like `eip` and `elb`.
I'm fully OK I will change any names I put in resources I created this time :smiley:

radeksimko commented:
> Then I just followed other resources which have elastic prefix like eip and elb.

I think you have a good point about EIP. I guess that changing the naming of existing resource may be painful, but I believe that `aws_elastic_ip` would be better in this case.

ELB is AFAIK commonly used abbreviation and also these are 3 words instead of 2, so the longer version would be IMO too long (`aws_elastic_load_balancer`). // correct me if I'm wrong

blacked commented:
I prefer match name with aws lib, like: `aws_elasticcache`

phinze commented:
+1 @blacked - it's a bit of an oddity, but there's only a single "C" in the service title "AWS ElastiCache", so technically the full name should be `aws_elasticache` and not `aws_elastic_cache` :+1:

tmtk75 commented:
`aws_elasticache` sounds good to me.
I changed resource names like `aws_elasticache`, `aws_elasticache_subnet_group` and `aws_elasticache_security_group`.

phinze commented:
@tmtk75 looks like this is getting close! think you could give this a rebase and squash and we'll give it a final review?

tmtk75 commented:
I rebased on the current head of master and squashed as https://github.com/tmtk75/terraform/commit/6750ea79b976d1701bfab55a177d15436afcf75a
Is this OK? Please take a look.

chrisferry commented:
I can't seem to launch a redis cluster.
* Error creating Elasticache: Cannot create a redis cluster with a NumCacheNodes parameter greater than 1.
more info:
http://pastie.org/pastes/10090257/text

tmtk75 commented:
@chrisferry Thank you for the feedback.
I was able to reproduce the below error if I changed the value of https://github.com/tmtk75/terraform/blob/6750ea79b976d1701bfab55a177d15436afcf75a/builtin/providers/aws/resource_aws_elasticache_test.go#L151 to 2.
```
Cannot create a redis cluster with a NumCacheNodes parameter greater than 1.
```

If my understanding is correct, actually we cannot set 2 or more if engine is `redis` according to this page.
http://docs.aws.amazon.com/cli/latest/reference/elasticache/create-cache-cluster.html

> For clusters running Redis, this value must be 1. For clusters running Memcached, this value must be between 1 and 20.

I don't use redis on AWS and I didn't know that.

tmtk75 commented:
FYI: At f4e8d7162994db249a367b8f1aa9e793e3634395, I fixed availability_zone of acceptance test from `ap-northeast-1a` to `us-west-2a` and I rebased on the current master again.

tabolario commented:
I'd love to see this one in! Lack of full support for ElastiCache resources has been one thing holding me back from Terraform so far.

phinze commented on builtin/providers/aws/resource_aws_elasticache.go:
```diff
@@ -0,0 +1,256 @@
+	_, sterr := stateConf.WaitForState()
+	if sterr != nil {
+		return fmt.Errorf("Error waiting for elasticache (%s) to be deletable: %s", d.Id(), sterr)
+	}
```
Wouldn't it make more sense to move this polling into the end of `resourceAwsElasticacheCreate` so that by the time we return from that we can be sure that the cluster is ready to go?

tmtk75 replied:
OK! It's really nice advice!
I fixed at https://github.com/tmtk75/terraform/commit/0b10004c29cf5967fd4dd6bc7eb38b60a0031b1a
and confirmed the acceptance test worked.

phinze commented on builtin/providers/aws/resource_aws_elasticache.go:
```diff
@@ -0,0 +1,256 @@
+				return nil, "", nil
+			}
+
+			log.Printf("[ERROT] CacheClusterStateRefreshFunc: %s", err)
```
Typo here.

tmtk75 replied:
Fixed at https://github.com/tmtk75/terraform/commit/89035efed5b4afb1307ef1e55025a38546f51a40

phinze commented on builtin/providers/aws/resource_aws_elasticache.go:
```diff
@@ -0,0 +1,256 @@
+
+func resourceAwsElasticacheUpdate(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
```
If the `Update` is just an empty func, let's just omit it from the resource definition.

radeksimko replied:
or maybe better implement the Update function? :wink: 

http://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_ModifyCacheCluster.html

![screen shot 2015-04-14 at 16 21 50](https://cloud.githubusercontent.com/assets/287584/7140285/672282ae-e2c2-11e4-9d27-3ee4c0c7e17d.png)

phinze replied:
Heh it's true we should eventually do that - I just didn't want to force a bunch of extra work on our friend @tmtk75 since he's put in so much already.

We've generally seen a pattern of (a) initial impl with `ForceNew` everywhere and no update, (b) follow on with update support.

But that's not saying I'd complain if the implementation showed up. :wink:

radeksimko replied:
Fair point, that's understandable, it can be the next step.

tmtk75 replied:
Fixed at https://github.com/tmtk75/terraform/commit/5c96957abcce8f1d0a982b15ce41553fd2c067d5

phinze commented:
Ok @tmtk75 I gave this a final review here - it's really looking great overall.

I made last round of comments inline - once we get those sorted we can land this thing! :airplane:

tmtk75 commented:
Hi, @phinze I'm sorry for late reply. I think I fixed three points you indicated and I've replied to your inline comments. Could you please take a look?

Actually now this pull request seems not to be merged to master branch, but I think it's easy to resolve the conflict at a glance. So I'm going to do conflict resolution if this review ends with OK.

mbrevoort commented:
Excited to get this one merged as well.

juniorplenty commented:
@phinze any update?

phinze commented:
@juniorplenty i'll make sure we get this in for 0.5 - we'll need to get it ported over to upstream aws-sdk-go

tmtk75 commented:
@phinze Do I have something to do now? Might I merge the master branch to this pull request resolving conflict?
Please let me know if you guys have.

phinze commented:
@tmtk75 if you want to give the conversion a shot that would be great - if you take a look at [@catsby's recent pull request history](https://github.com/hashicorp/terraform/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Acatsby+is%3Aclosed+convert) you'll see the sort of changes that need to be made to move from `hashicorp/aws-sdk-go` over to the upstream `awslabs` repo.

tmtk75 commented:
@phinze Hi, I think I was able to fix as you say so that it uses `awslabs`. Please take a look.

- Just merged the current master once at 65ccc99656839edf73fad81ad369045734f975fd
- Moved to `awslabs` at 700e08f850aea9f1fb208e5136b873def4664e57
- Fixed cidr_block of vpc in the acceptance test at 763a9ae30dee7a71fe9d2408e6495fd319f358ea because it failed due to https://github.com/tmtk75/terraform/blob/master/helper/resource/testing.go#L248 for `aws_vpc` resource.

Certainly I confirmed the acceptance test was passed and I'm ready for rebase and squash again if it's needed.

phinze commented on builtin/providers/aws/resource_aws_elasticache.go:
```diff
@@ -0,0 +1,245 @@
+}
+
+func resourceAwsElasticacheCreate(d *schema.ResourceData, meta interface{}) error {
+	elasticacheconn := meta.(*AWSClient).elasticacheconn
```
This is a minor style thing, but we've been trying to call local client vars `conn` everywhere, [example](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_instance.go#L309).

phinze commented on builtin/providers/aws/resource_aws_elasticache.go:
```diff
@@ -0,0 +1,245 @@
+	//     Can only delete cache clusters with state in:
+	//     available, failed, incompatible-parameters, incompatible-network, restore-failed
+	// So we cannot delete clusters in `creating` state.
+	// We need to wait for states which we can delete.
```
This comment is a bit stale now since we moved this to Create. The below block is basically summed up in the debug log line below "waiting for state to become available."

phinze commented on builtin/providers/aws/resource_aws_elasticache.go:
```diff
@@ -0,0 +1,245 @@
+	log.Printf("[DEBUG] Waiting for state to become available: %v", d.Id())
+	_, sterr := stateConf.WaitForState()
+	if sterr != nil {
+		return fmt.Errorf("Error waiting for elasticache (%s) to be deletable: %s", d.Id(), sterr)
```
This message needs an update now that this lives in `Create` "to be deletable" -> "to be created"

phinze commented:
Amazing amazing work here @tmtk75!

I commented on a final couple of very minor things I noticed (which you're welcome to wait until after merge to fix if you want.)

Then if you could give this a squash, I'll press the big green button. :grinning:

tmtk75 commented:
@phinze Thank you for your quick reply!
I fixed as your comments at 9613deac3c137378efa00df35a9c5af2a45b676b and then squashed and pushed f3d9a287e1bdb4720d540b5961b7bfdece47ab9f.
Hoping this commit would be fine to be merged :smiley:

phinze commented:
Looks great. Time to land. :rocket: :guitar:

tmtk75 commented:
I'm very excited :tada: Thanks a lot for your reviews, advices and feedbacks, @phinze and everyone else!

ketzacoatl commented:
Thank you for your work on this @tmtk75!! Very appreciated :)
